### PR TITLE
Drag'n'Drop implementation for DisplayButtons

### PR DESCRIPTION
**Pull request by jgersti**
## Overview
This PR implements ~~rudamentary~~ drag'n'drop support to the `DisplayButton`s in the Options.
Feature and Design is descripted in some [sketches](https://drive.google.com/drive/folders/0BzaKJginnJcgM3dHWUdDQlNRZGM?usp=sharing).

The implemented features make several `DisplayButton` controls in their current form obsolete.
Chiefly the side control of grouped  buttons (up, down, ungroup) which also distinguish grouped buttons from ungrouped one. 
(see https://www.wowace.com/projects/weakauras-2/issues/790 )

### Issues
-    ~~If the `DisplayButtonList` is *not* scrolled to the top, the `DisplayButton` is not __*visually*__ dragged, but dropping it still works at the __cursor__ location.~~
-    ~~If the `DisplayButtonList` is being scrolled while dragging, the dragged element is shown in its original poistion.  Dropping it still works at the __cursor__ location.~~
-    ~~`OnDragStop` is not called correctly when entering combat while dragging a `DisplayButton`. This results in a hidden element until reloading.~~
-    Dropping a element below the list of `DisplayButton` into the scroll frame or onto a category rider undos the drag action. This behavior might be unexpexted.

### CODE REVIEW

**Comment by Stanzilla** [comment]
Oh neat! I will take a look once we shipped https://github.com/WeakAuras/WeakAuras2/pull/135

**Comment by InfusOnWoW** [comment]
We are definately interested in having that in WeakAuras. If you have any questions or need any help feel free to ask. We do hang out in the #weakauras irc channel on freenode.

**Comment by jgersti** [comment]
I did investigate the scroll issues during the week and found that the scroll issues are caused by the fact that the function `SortDisplayButton` repopulates the the scroll frame and trigger an recalculation of the layout.
My first attempt at fixing the issues was to add a more general implemenation for filtering into `SortDisplayButton` and use a exclude filter on the dragged button. This made the buttons dragable regardless of the scroll state or actual scrolling while dragging but no drop action was ever fired anymore unless the frame was scrolled to the top..

I also i managed to corrupt a group by inserting a button twice and while fixing my saved variables by hand i noticed i polluted the then with my drag state marker.
The last comment lets me to the question where i can savely put the variable so it is not pulled into the saved variables.

**Comment by InfusOnWoW** [comment]
You asked why DoLayout is called on each scroll, and the reason is this: https://github.com/WeakAuras/WeakAuras2/commit/768f9e66b044d3e0ace0e15897a28479c4fbbaaa

**Comment by jgersti** [comment]
I fixed the scroll issues. 
That leaves open some cosmetic changes, like for exampe a visual hint where the dragged element would be inserted, and more feature like moving buttons into or between groups.
Especially for something like a visual drop hint i would need some help with the visualization and where to put the controlling variables.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 1]
```diff
@@ -436,57 +428,56 @@ local methods = {
...
+            local scale = UIParent:GetScale()
 
-            self.frame:StartMoving()
-            self.frame:SetUserPlaced(false)
+            frame.isDragged = true
+            frame:StartMoving()
+            frame:ClearAllPoints()
+            frame:SetPoint("Center", UIParent, "BOTTOMLEFT", x/scale, y/scale)
+
+            WeakAuras:SortDisplayButtons() -- find a better way to trigger a DoLayout call on the scroll frame
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: needs to be WeakAuras.SortDisplayButtons

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 2]
```diff
@@ -63,10 +79,10 @@ local methods = {
             elseif(IsShiftKeyDown()) then
                 local editbox = GetCurrentKeyBoardFocus();
                 if(editbox) then
-				    if (not fullName) then
-				      local name, realm = UnitFullName("player")
-					  fullName = name.."-"..realm
-					end
+                    if (not fullName) then
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Can you split up whitespace clean ups into separate commits? (The formating is funky in places, and I'm okay with correcting that. But it makes it harder to review patches.)

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 3]
```diff
@@ -420,6 +436,60 @@ local methods = {
...
+                if i then
+                    if IsOverDisplay(parent) then
+                        tremove(children, i)
+                        tinsert(children, 1, id)
+                        WeakAuras.Add(parentData);
+                        UpdateGroupOrder(children)
+                    else
+                        local j
+                        for key, value in pairs(children) do
+                            if key ~= i and IsOverDisplay(value) then
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: I think it would be better if the top-half of each button would insert before that child, and the bottom half after that child.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 4]
```diff
@@ -420,6 +436,60 @@ local methods = {
             end
         end
 
+        function self.callbacks.OnDragStart(_, _)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Can you make this function hide various bits of the frame while dragging? E.g. the visibility icon, the ungroup, the move up/down icons.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 5]
```diff
@@ -420,6 +436,60 @@ local methods = {
...
+                            UpdateGroupOrder(children)
+                        end
+                    end
+                else
+                    error("Display thinks it is a member of a group which does not control it")
+                end
+                WeakAuras.SortDisplayButtons()
+                WeakAuras.UpdateDisplayButton(parentData)
+
+                data.dragging = false
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Looking at this code, it looks to me that adding dragging between groups wouldn't be hard to add, right?

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 6]
```diff
@@ -436,57 +428,56 @@ local methods = {
             end
         end
 
-        function self.callbacks.OnDragStart(_, _)
+        function self.callbacks.OnDragStart(frame, _)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: You don't need the "_" parameter, lua ignores if a function takes less arguments than the caller provided.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 1 above)
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Just add a method to the WeakAuras table that does that.

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 7]
```diff
@@ -363,34 +363,36 @@ AceGUI:RegisterLayout("AbsoluteList", function(content, children)
 end);
 
 AceGUI:RegisterLayout("ButtonsScrollLayout", function(content, children)
-  local yOffset = 0;
-  local scrollTop, scrollBottom = content.obj:GetScrollPos();
+  local yOffset = 0
+  local scrollTop, scrollBottom = content.obj:GetScrollPos()
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: Can you keep the ";"? It's optional, but most of the code base has trailing ;

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 8]
```diff
@@ -436,57 +428,56 @@ local methods = {
...
+            frame:ClearAllPoints()
+            frame:SetPoint("Center", UIParent, "BOTTOMLEFT", x/scale, y/scale)
+
+            WeakAuras:SortDisplayButtons() -- find a better way to trigger a DoLayout call on the scroll frame
         end
 
-        function self.callbacks.OnDragStop(_)
-            if data.dragging then
-                self.frame:StopMovingOrSizing()
+        function self.callbacks.OnDragStop(frame)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: As to the visualization, there seems to be no callback for dragMove. So the way I would do that.

In OnDragStart, register a OnUpdate method on frame (But check if no other code does that, so that there aren't any conflits.) Unregister that method in OnDragStop.

In that OnUpdate handler, you need to basically figure out which position the frame would be added if the drop stopped at that moment. That's essentially the same code you already have.

Also you need a frame that represents the drop indicator, so add a method to this file: dropIndicator() which lazily creates a frame with a texture child.

In the OnUpdate handler you position the frame relative to the button by using SetPoint.

That doesn't sound to complicated to me.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 9]
```diff
@@ -436,57 +428,56 @@ local methods = {
...
-                                break
-                            end
-                        end
+                        local j, v
+                        repeat
+                            j, v = next(children, j)
+                        until not j or j ~= i and IsOverDisplay(v)
                         if j then
                             tremove(children, i)
                             tinsert(children, j+1, id)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: This looks wrong. Suppose the user is dropping i = 2 onto j=3, then removing i = 2, makes j move up a place.
Compare that to i = 3 onto j = 2, where removing i doesn't affect j.

**Comment by InfusOnWoW** [comment]
Also it would be vital to have a indicator that drag and drop is now supported.

I think the canonical drag handle looks like this: http://designingwebinterfaces.com/blog/wp-content/uploads/2009/07/gmail_drag-handle.png 

Also somewhat loosely related is this ticket:
https://www.wowace.com/projects/weakauras-2/issues/790
That might be worth considering if you add a drag handle

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 1 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: thta happen when you retype your code from the live version into the repo. ;)

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 2 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: i can seperate it. removing it might be difficult since i guess this where tabs previously.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 6 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Its just for completness sake. If you want i can remove it.

File: WeakAurasOptions/WeakAurasOptions.lua (hunk 7 above)
Comment (jgersti) [inline WeakAurasOptions/WeakAurasOptions.lua]: Sure.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 1 above)
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: We recently changed the folder structure in such a way that you can actually use a git checkout. (You'll be missing translations, but whatever)

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 6 above)
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Standard lua is to only mention those parameters that you use.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 8 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: The logic is not really the problem. 
I just have no idea if a frame anchored to the hovered element would survive a scroll action
and how to create appropriate a frame/texture.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 9 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: You should be right.  The following should give the right behaviour:

```lua
tremove(children, i)
if i < j then
   tinsert(children, j, id)
else
    tinsert(children,j+1, id)
end

```

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 1 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: I tried not to add any functions into the adon namespace :)

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 8 above)
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: You anchor to the hoovered element. Also you are updating on each frame anyway. As to how create a frame with a texture. That's something like this:

  local dragIndicator = CreateFrame("FRAME", "WeakAurasDragIndicator");
  dragIndicator:SetHeight(10); -- TODO depends
  dragIndicator:SetFrameStrata("FULLSCREEN"); -- above settings dialog
  
  local texture = dragIndicator:CreateTexture(nil, "BACKGROUND");
  texture:SetAllPoints(dragIndicator);
  texture:SetTexture("a path");

and then in once you have the frame to which you want to anchor:
dragIndicator:ClearAllPoints();
dragIndicator:SetPoint("TOPLEFT", anchor, "BOTTOMLEFT", xoffset, yoffset);
dragIndicator:SetPoint("TOPRIGHT", anchor, "BOTTOMRIGHT", xoffset, yoffset);

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 9 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: I will test both ways later to see which on behaves the right way.
I must admit i only tested with a group with three children since i did not want to corrupt anything complicated.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 1 above)
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: It's not worse thatn having WeakAuras.SortDisplayButtons there

**Comment by Stanzilla** [comment]
Got this error during testing:

```
5x WeakAurasOptions\WeakAurasOptions-@project-version@.lua:8989: Usage: WeakAurasFilterInput:SetText("text")
[C]: in function `SetText'
WeakAurasOptions\WeakAurasOptions-@project-version@.lua:8989: in function `SortDisplayButtons'
...eGUI-Widgets\AceGUIWidget-WeakAurasDisplayButton.lua:442: in function <...eGUI-Widgets\AceGUIWidget-WeakAurasDisplayButton.lua:431>

Locals:
(*temporary) = WeakAurasFilterInput {
 0 = <userdata>
 Right = <unnamed> {
 }
 Left = <unnamed> {
 }
 Middle = <unnamed> {
 }
}
(*temporary) = <table> {
 SetAuraVisibility = <function> defined @WeakAuras\BuffTrigger.lua:288
 ShowSpellIDDialog = <function> defined @WeakAurasOptions\WeakAurasOptions.lua:9420
 GetRuneCooldown = <function> defined @WeakAuras\GenericTrigger.lua:1238
 frame_strata_types = <table> {
 }
 DisplayToString = <function> defined @WeakAuras\Transmission.lua:451
 combat_event_type = <table> {
 }
 regions = <table> {
 }
 CheckItemSlotCooldowns = <function> defined @WeakAuras\GenericTrigger.lua:1542
 duration_types = <table> {
 }
 EnsureClone = <function> defined @WeakAuras\WeakAuras.lua:2060
 debug = <function> defined @WeakAuras\WeakAuras.lua:222
 GetIconFromSpellCache = <function> defined @WeakAurasOptions\WeakAurasOptions.lua:434
 CanShowNameInfo = <function> defined @WeakAuras\WeakAuras.lua:2690
 tutorials = <table> {
 }
 pAdd = <function> defined @WeakAuras\WeakAuras.lua:1814
 Rename = <function> defined @WeakAuras\WeakAuras.lua:1221
 IsRegisteredForCustomTextUpdates = <function> defined @WeakAuras\WeakAuras.lua:2985
 LayoutDisplayButtons = <function> defined @WeakAurasOptions\WeakAurasOptions.lua:1464
 transmitCache = <table> {
 }
 CreateEncounterTable = <function> defined @WeakAuras\WeakAuras.lua:829
 event_types = <table> {
 }
 TenchInit = <function> defined @WeakAuras\GenericTrigger.lua:2113
 regionTypes = <table> {
 }
 miss_types = <table> {
 }
 GetGCDInfo = <function> defined @WeakAuras\GenericTrigger.lua:1274
 faction_group = <table> {
 }
 OpenOptions = <function> defined @WeakAuras\WeakAuras.lua:63
 autoeventend_types = <table> {
 }
 CanGroupShowWithZero = <function> defined @WeakAuras\WeakAuras.lua:2618
 talent_types = <table> {
 }
 GetSpellCooldown = <function> defined @WeakAuras\GenericTrigger.lua:1246
 LoadFunction = <function> defined @WeakAuras\WeakAuras.lua:420
 circular_group_constant_factor_types = <table> {
 }
 GetAuraTooltipInfo = <function> defined @WeakAuras\WeakAuras.lua:2804
 events = <table> {
 }
 AddManyFromAddons = <function> defined @WeakAuras\WeakAuras.lua:1349
 WatchForMounts = <function> defined @WeakAuras\GenericTrigger.lua:2181
 spec_types_3 = <table> {
 }
 OpenModelPick = <function> defined @WeakAurasOptions\WeakAurasOptions.lua:9265
 RegisterLoadEvents = <function> defined @WeakAuras\WeakAuras.lua:1092
 blend_types = <table> {
 }
 ImportString = <function> defined @WeakAuras\Transmission.lua:1073
 GetPolarCoordinates = <function> defined @WeakAuras\RegionTypes\dynamicgroup.lua:38
 GetDynamicIconCache = <function> defined @WeakAuras\WeakAuras.lua:3113
 custom_trigger_types = <table> {
 }
 group_types = <table> {
 }
 CollapseAllClones = <function> defined @WeakAuras\WeakAuras.lua:2086
 partyUnits = <table> {
 }
 IsSpellInRange = <function> defined @WeakAuras\Prototypes.lua:23
 SetDynamicIconCache = <function> defined @WeakAuras\WeakAuras.lua:3108
 text_rotate_types = <table> {
 }
 MultipleDisplayTooltipDesc = <function> defined @WeakAurasOptions\WeakAurasOptions.lua:71
 UpdateMouseoverTooltip = <function> defined @WeakAuras\WeakAuras.lua:2754
 spec_types = <table> {
 }
 item_slot_types = <table> {
 }
 RegisterCustomTextUpdates = <function> defined @WeakAuras\WeakAuras.lua:2976
 anim_color_types = <table> {
 }
 loaded = <table> {
 }
 AddOption = <function> defined @WeakAurasOptions\WeakAurasOptions.lua:2144
 Convert = <function> defined @WeakAuras\WeakAuras.lua:1287
```

**Comment by Stanzilla** [comment]
Got some corruption here while playing around with it: https://gfycat.com/ContentParchedArgentinehornedfrog

The error that happens at the end is

```
3x ...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0-61.lua:1451: Usage: GameTooltip:SetText("text" [, color, wrap])
[C]: in function `SetText'
...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0-61.lua:1451: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:1421>
(tail call): ?
[C]: ?
[string "safecall Dispatcher[4]"]:9: in function <[string "safecall Dispatcher[4]"]:5>
(tail call): ?
Ace3\AceGUI-3.0\AceGUI-3.0-34.lua:314: in function `Fire'
...Ace3\AceGUI-3.0-34\widgets\AceGUIContainer-TabGroup.lua:73: in function <...Ace3\AceGUI-3.0\widgets\AceGUIContainer-TabGroup.lua:71>

Locals:
nil
```

**Comment by jgersti** [comment]
I am not sure the error is related to any changes i made directly. i got an a `SetText` error after i replaced my local files with the latest release version from curse.

**Comment by jgersti** [comment]
I reformated the commits to cleanup or seperate some (unwanted) changes.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 2 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: seperated into  d0d66a2

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 9 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: fixed in f5d7575

**Comment by jgersti** [comment]
Since the rebase hid some unaddressed comments i quote them again

> Can you make this function hide various bits of the frame while dragging? E.g. the visibility icon, the ungroup, the move up/down icons.

> I think it would be better if the top-half of each button would insert before that child, and the bottom half after that child.

**Comment by jgersti** [comment]
> Can you make this function hide various bits of the frame while dragging? E.g. the visibility icon, the ungroup, the move up/down icons.
see 4601e24

**Comment by jgersti** [comment]
3d948ae together f3d5e4a should hide the tooltip and menu while dragging

**Comment by jgersti** [comment]
e11cb2b reverts f3d5e4a

**Comment by jgersti** [comment]
I implemented a drop indicator yesterday and will push the current state later today.
That said i am not satisfied how it looks.
I would appreciate some help so the look and feel is richtig.

**Comment by Stanzilla** [comment]
I quite like the concept of a glowing line to indicate the position it will be dropped into, imagine this: https://gfycat.com/CarelessSinfulBluebird just vertically

**Comment by InfusOnWoW** [comment]
Also, I'm not sure how much Stanzilla dived into the code, but for me it would be easier if you squashed your changes. (That is all the ones related to this feature.) As there was some back and forth between a few commits.

If you are struggling with the drop indicator, finish everything else and we can polish that ourselves. This is alread a good patch.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 10]
```diff
@@ -44,6 +44,35 @@ local function Show_Long_Tooltip(owner, description)
     GameTooltip:Show();
 end
 
+local function IsOverDisplay(display)
+    if not display then return nil end
+
+    local button = WeakAuras.GetDisplayButton(display)
+    return MouseIsOver(button.frame)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: So if I read that correctly, http://wowwiki.wikia.com/wiki/API_MouseIsOver supports a top and bottom offset.

I still think that it should behave such that hoovering over the top half of a buttom should insert above it, and hoovering over the bottom half below that aura. 

That should be implementable by using half a width as bottom and topoffset (no idea on the sign), in such a way that the button steals the a half of the next button.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 11]
```diff
@@ -44,6 +44,35 @@ local function Show_Long_Tooltip(owner, description)
...
+end
+
+local function Show_DropIndicator(dataDragging)
+    local data, indicator = dataDragging, WeakAuras.DropIndicator()
+    if data then
+        local buttonList = WeakAuras.displayButtons
+        local id, button
+        repeat
+            id, button = next(buttonList, id)
+        until not id or id ~= data.id and button:IsEnabled() and button.frame:IsMouseOver()
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: Obviously would need to use this here too.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 12]
```diff
@@ -420,6 +449,67 @@ local methods = {
...
+                self.downgroup:Show()
+                self.loaded:Show()
+                self.ungroup:Show()
+                self.upgroup:Show()
+                self.view:Show()
+
+                local id, parent = data.id, data.parent
+                local parentData = WeakAuras.GetData(parent)
+                local children = parentData.controlledChildren
+
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: I'd prefer if this code would handle inter group dragging.

Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: I need to map out the logic for that.  
right now i just check that there is a parent which is the easy way :smile:  
In my eyes this should include to be able to put a ungrouped button into a group and vice versa.
This means there needs to be a visual hint (and the logic) what operation would performed on dropping.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 10 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: I wanted to defer that because i thought that would be complicated.
I realised by now that designing lua code is different from C++ :laugh:

**Comment by jgersti** [comment]
I did some sketches during my lunch break concernig some visualization and behaviour (with https://www.wowace.com/projects/weakauras-2/issues/790 in mind).
[Link](https://drive.google.com/drive/folders/0BzaKJginnJcgM3dHWUdDQlNRZGM?usp=sharing)

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 13]
```diff
@@ -673,6 +768,43 @@ local methods = {
...
+                    if self.elapsed > 0.1 then
+                        Show_DropIndicator(draggingData)
+                        self.elapsed = 0
+                    end
+                end)
+
+                self.frame.isDragged = true
+                self.frame:StartMoving()
+                self.frame:ClearAllPoints()
+                self.frame:SetPoint("Center", UIParent, "BOTTOMLEFT", x/scale, y/scale)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: what you might try here, is to only move the frame vertically. That is, set the x position not based on the mouse cursor, but based on the list's position. No idea how much code that is, but I think that might look better visually.

**Comment by InfusOnWoW** [comment]
Those pictures look nicely done and as far as I can see all the interactions are correct.

Now, there's ton of details. I'm not sure whether the right approach would be to "finish" the feature without implementing all details. (E.g. my suggestion locking the x axis, or a "perfect" drop indicator. ) Merge that, so that we already have something. And you can continue working on refinements piece by piece.

@Stanzilla what do you think?

**Comment by Stanzilla** [comment]
I almost agree with @InfusOnWoW and completely with your mockups. I want to release the anchors feature with the next release and would merge a first implementation of your drag n drop afterwards. I really like the idea of the arrow in addition to the glowing line as the drop indicator, I think that will be a very good visual aid and I think the indicator should be part of the first implementation since it will be almost guaranteed that people won't understand how this feature works if it is not.

That said, either I'm blind or your current implementation of the drop indicator is not working for me. 
See this gif https://gfycat.com/IncrediblePerfectGermanspaniel

At the end you can see what a first time user will have problems with when trying it, he will try to move CP1 above CP2 and fail to do so. That's basically why we need the indicator. If I just fucked up my local checkout and this is actually working in your code, this ofc is invalid and I'm sorry. 

So in conclusion, I really look forward to this feature and love your concepts and will definitely merge it after we ship anchors and have a working drop indicator! 👍 

A good follow up would be a way to drag single auras into a group for example.

**Comment by jgersti** [comment]
So most of the stuff detailed in my sketches is implemented.
Now i need some people to test for corner case i might have missed and to test usablity and how intuitive the feature is.

Most visual aspect will need some further refinement which i am not too experienced with.

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 14]
```diff
@@ -44,6 +48,221 @@ local function Show_Long_Tooltip(owner, description)
...
+                    return function(_source, _target)
+                            Actions["Move"](_source, _target, area=="TOP")
+                        end,
+                        Icons["Move"]
+                else
+                    return function(_source, _target)
+                            Actions["Ungroup"](_source)
+                            Actions["Group"](_source, _target.data.parent, _target, area == "TOP")
+                        end,
+                        Icons["Group"],
```
Comment (Stanzilla) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: the komma here will error

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua [hunk 15]
```diff
@@ -44,6 +48,221 @@ local function Show_Long_Tooltip(owner, description)
...
+    end,
+}
+
+local Icons = {
+    ["Group"] = "Interface\\GossipFrame\\TrainerGossipIcon",
+    ["Ungroup"] = "Interface\\GossipFrame\\UnlearnGossipIcon",
+    ["Move"] = nil
+}
+
+local function GetAction(target, area, source)
```
Comment (Stanzilla) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: you haven't declared _source and _target at the top of the function it will be invalid

Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: these are the argument of the function i am returning

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 14 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: fixed in  c95c071

File: WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua (hunk 15 above)
Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: sry you are right i forgot to remove the arguments

Comment (jgersti) [inline WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasDisplayButton.lua]: fixed in 9922dd3

**Comment by Stanzilla** [comment]
Alright, do you think this is ready to get more testing in the alpha versions?

**Comment by jgersti** [comment]
I think the feature is ready for testing.
The only unsatisfactory part is the visualization maybe the last issue I
added. But these need more feedback. :smile:

Am 23.01.2017 2:04 vorm. schrieb "Benjamin Staneck" <
notifications@github.com>:

**Comment by Stanzilla** [comment]
Yeah I figured, the icon for in/out of group moving looked like a placeholder to me :)
Can you rebase onto master then, please?
